lib/liquid/c.rb

```diff
@@ -14,22 +20,38 @@ def self.new(source, line_numbers = false, line_number: nil, for_liquid_tag: fal
   end
 end
 
-Liquid::BlockBody.class_eval do
-  alias_method :ruby_parse, :parse
+Liquid::ParseContext.class_eval do
+  # liquid-c now generates a different parse tree, which can cause incompatibilities issues
+  # where there is internal coupling.  Allow that to be disabled for now by passing a
+  # :disable_liquid_c_nodes option to Template.parse.
+  def disable_liquid_c_nodes
+    @disable_liquid_c_nodes ||= @template_options.fetch(:disable_liquid_c_nodes, !Liquid::C.enabled)
+  end
+end
 
-  def parse(tokens, options)
-    if Liquid::C.enabled && !options[:profile]
-      c_parse(tokens, options) { |t, m| yield t, m }
-    else
-      ruby_parse(tokens, options) { |t, m| yield t, m }
+[Liquid::Block, Liquid::Document].each do |klass|
+  klass.class_eval do
+    private
+
+    alias_method :ruby_new_body, :new_body
+
+    def new_body
+      if parse_context.disable_liquid_c_nodes || parse_context[:profile]
```

Comment:
Does profiling disables this only because instrumentation is missing from `C::BlockBody` at the moment?

Reply:
Liquid's template render profiling support already wasn't supported in liquid-c.  I just chose to keep it that way.  Note that this is different from stackprof, which we typically use for profiling at Shopify.  I would like to revisit profiling in the future, but not in a way that ties us to calling a `render_node` ruby method that requires a node object to be passed to it.